File: cpp-gcc/test/NameNormalizerTest.cpp
#include "gmock/gmock.h"

#include "NameNormalizer.h"

using namespace testing;
using namespace std;
using namespace name_util;

TEST(ANameNormalizer, ReturnsAnEmptyStringWhenGivenSame) {
    ASSERT_THAT(normalize(""), StrEq(""));
}

TEST(ANameNormalizer, SimplyReturnsASingleWordName) {
    ASSERT_THAT(normalize("Plato"), StrEq("Plato"));
}

TEST(ANameNormalizer, SwapsFirstAndLastNames) {
    ASSERT_THAT(normalize("Haruki Murakami"), StrEq("Murakami, Haruki"));
}

TEST(ANameNormalizer, TrimsWhitespace) {
    ASSERT_THAT(normalize("  Big Boi   "), StrEq("Boi, Big"));
}

TEST(ANameNormalizer, ReplacesMiddleNameWithInitial) {
    ASSERT_THAT(normalize("Henry David Thoreau"), StrEq("Thoreau, Henry D."));
}

TEST(ANameNormalizer, DoesNotInitializeOneLetterMiddleName) {
    ASSERT_THAT(normalize("Harry S Truman"), StrEq("Truman, Harry S"));
}

TEST(ANameNormalizer, InitializesEachOfMultipleMiddleNames) {
    ASSERT_THAT(normalize("Julia Scarlett Elizabeth Louis-Dreyfus"), StrEq("Louis-Dreyfus, Julia S. E."));
}

TEST(ANameNormalizer, AppendsSuffixesToEnd) {
    ASSERT_THAT(normalize("Martin Luther King, Jr."), StrEq("King, Martin L., Jr."));
}

TEST(ANameNormalizer, ThrowsWhenNameContainsTwoCommas) {
    ASSERT_THROW(normalize("Thurston, Howell, III"), InvalidNameException);
}

File: elixir/test/name_normalizer_test.exs
defmodule NameNameNormalizerTest do
  use ExUnit.Case

  test "returns empty string when empty" do
    assert NameNormalizer.normalize("") == {:ok, ""}
  end

  test "returns single word name" do
    assert NameNormalizer.normalize("Plato") == {:ok, "Plato"}
  end

  @tag :skip
  test "swaps first and last names" do
    assert NameNormalizer.normalize("Haruki Murakami") == {:ok, "Murakami, Haruki"}
  end

  @tag :skip
  test "trims leading and trailing whitespace" do
    assert NameNormalizer.normalize("  Big Boi   ") == {:ok, "Boi, Big"}
  end

  @tag :skip
  test "initializes middle name" do
    assert NameNormalizer.normalize("Henry David Thoreau") == {:ok, "Thoreau, Henry D."}
  end

  @tag :skip
  test "does not initialize one letter middle name" do
    assert NameNormalizer.normalize("Harry S Truman") == {:ok, "Truman, Harry S"}
  end

  @tag :skip
  test "initializes each of multiple middle names" do
    assert NameNormalizer.normalize("Julia Scarlett Elizabeth Louis-Dreyfus") ==
             {:ok, "Louis-Dreyfus, Julia S. E."}
  end

  @tag :skip
  test "appends suffixes to end" do
    assert NameNormalizer.normalize("Martin Luther King, Jr.") == {:ok, "King, Martin L., Jr."}
  end

  @tag :skip
  test "returns an error when name contains two commas" do
    assert NameNormalizer.normalize("Thurston, Howell, III") == {:error, "Too many commas"}
  end
end
